updater/internal/file/file.go

```diff
@@ -24,9 +25,80 @@ import (
-// If the file does not exist, it is created. If the file does exist, it is truncated before writing.
-func CopyFile(logger *zap.Logger, pathIn, pathOut string, overwrite bool, useInFilePermBackup bool) error {
+// CopyFileOverwrite copies the file from pathIn to pathOut.
+// The output file is created if it does not exist.
+// If the output file does exist, it is removed, then written from the input file, preserving the output file's mode.
+func CopyFileOverwrite(logger *zap.Logger, pathIn, pathOut string) error {
+	fileMode := fs.FileMode(0600)
+	pathOutClean := filepath.Clean(pathOut)
+
+	// Try to save existing file's permissions
+	outFileInfo, _ := os.Stat(pathOutClean)
+	if outFileInfo != nil {
+		fileMode = outFileInfo.Mode()
+	}
+
+	pathInClean := filepath.Clean(pathIn)
+	// If the input file cannot be opened for some reason, do NOT delete the file
+	if _, err := os.Stat(pathInClean); err != nil {
+		return fmt.Errorf("failed to stat input file: %w", err)
+	}
+
+	// Remove old file to prevent issues with mac
+	if err := os.Remove(pathOutClean); err != nil {
+		logger.Debug("Failed to remove output file", zap.Error(err))
+	}
+
+	return copyFileInternal(logger, pathIn, pathOut, os.O_CREATE|os.O_WRONLY, fileMode)
+}
+
+// CopyFileNoOverwrite copies the file from pathIn to pathOut, preserving the input file's mode.
+// If the output file already exists, this function returns an error.
+func CopyFileNoOverwrite(logger *zap.Logger, pathIn, pathOut string) error {
+	pathInClean := filepath.Clean(pathIn)
+
+	// Use the new file's permissions and fail if there's an issue (want to fail for backup)
+	inFileInfo, err := os.Stat(pathInClean)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve fileinfo for input file: %w", err)
+	}
+
+	// the os.O_EXCL flag will make OpenFile error if the file already exists
+	return copyFileInternal(logger, pathIn, pathOut, os.O_EXCL|os.O_CREATE|os.O_WRONLY, inFileInfo.Mode())
+}
+
+// CopyFileRollback copies the file to the file from pathIn to pathOut, preserving the input file's mode if possible
+// Used to perform a rollback
+func CopyFileRollback(logger *zap.Logger, pathIn, pathOut string) error {
+	// Default to 0600 if we can't determine the input file's mode
+	fileMode := fs.FileMode(0600)
+	pathInClean := filepath.Clean(pathIn)
+	// Use the backup file's permissions as a backup and don't fail on error (best chance for rollback)
+	inFileInfo, err := os.Stat(pathInClean)
+	switch {
+	case errors.Is(err, os.ErrNotExist):
+		return fmt.Errorf("input file does not exist: %w", err)
+	case err != nil:
+		// TODO: Should this actually error out?
+		// If we can't STAT the input file, what are the chances we can open for reading?
+		// we are going to delete the file, so we might be better off bailing here and leaving the
+		// newer file instead of deleting it and failing to copy the input file
+		logger.Error("failed to retrieve fileinfo for input file", zap.Error(err))
```

Comment:
I'm not sure if continuing in this case is actually OK; I'm thinking it might make sense to just error out here instead for the reasons I have in this comment.

Any thoughts?

Reply:
you're most likely right? but i don't know about just erroring out completely. we should let things continue just in case and give it the best chance to roll back as completely as possible

Reply:
overall this PR is an improvement though. 👍

Reply:
Added comment about why we continue here.